### PR TITLE
feat: apdex skip tests on demand and cleaning

### PR DESCRIPTION
**Pull request by latin-panda**
# Description

This PR includes the following, for more details, see the comments below in the code. 

- New settings: 
   - Relaunching the app is now optional, and we can specify it on any page. 
   - Search is now defined at page level, so we can search in contact list or report list or any other page in the future.
   - Now, we define the path of the APK file in the computer. Not hardcoded anymore. 
   - Option to start faster by skipping the APK installation and the login: `skip.login` and `capabilities.noReset`, especially useful when building the tests for a new config and want to test if they run fine. 
   - Skip test cases from the settings file. 
- Small refactors
  - one of them is to have a `clickButton` utility function that uses our wait mechanism with automatic retries.  
  - Renaming to follow conventions
    -  `performance.page.js` file to `targets.page.js`
    - `test.apdex.js` to `test.spec.js`
    -  `searchPatient` to `searchContact`
    - `contact-list`, `messages`, `reports` and other labels to follow same convention: `contactList`, `messageList`, `reportList`, etc.
- Separated task load and task submit test cases
- Added report search test case. 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:apdex-skip-tests-on-demand/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:apdex-skip-tests-on-demand/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:apdex-skip-tests-on-demand/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/page-objects/apdex/targets.page.js
Comment (latin-panda) [inline tests/page-objects/apdex/targets.page.js:1]: I renamed this file from `performance.page.js` to `targets.page.js` because in Medic we always use `targets` term when talking about this page ([see documentation](https://docs.communityhealthtoolkit.org/apps/features/targets/))
`Performance` is the translation used by a partner, and they can use any word such as `Analytics`.

File: tests/page-objects/apdex/contacts.page.js
Comment (latin-panda) [inline tests/page-objects/apdex/contacts.page.js:42]: Renaming from `searchPatient` to `searchContact`, since technically, anyone can configure the tests to search by any household, facility, or patient.

File: tests/page-objects/apdex/messages.page.js
Comment (latin-panda) [inline tests/page-objects/apdex/messages.page.js:6]: Name to follow camelcase and same line of naming as `contactList`

File: tests/page-objects/apdex/load.page.js
Comment (latin-panda) [inline tests/page-objects/apdex/load.page.js:15]: Skips instance setup. When the app is already configured, it won't show the instance setup page.

File: tests/page-objects/apdex/page.js
Comment (latin-panda) [inline tests/page-objects/apdex/page.js:120]: New setting. 
Relaunching the app is now optional and we can specify it on any page.

File: tests/page-objects/apdex/page.js
Comment (latin-panda) [inline tests/page-objects/apdex/page.js:195]: New setting.
Now we can search in the contact list or the report list.

File: tests/page-objects/apdex/page.js
Comment (latin-panda) [inline tests/page-objects/apdex/page.js:201]: @ralfudx @tatilepizs, I need help with this one. I didn't find how to improve these selectors.
Can you please help me?

File: tests/page-objects/apdex/tasks.page.js
Comment (latin-panda) [inline tests/page-objects/apdex/tasks.page.js:11]: The form has a `navigation` property, so this is no necessary.

File: tests/performance/apdex-score/apk/CHTAndroid.apk
Comment (latin-panda) [inline tests/performance/apdex-score/apk/CHTAndroid.apk:1]: New setting. 
We shouldn't hardcore the APK. 
With this new setting, you can easily specify the APK path on your computer, streamlining your testing process and saving time. 
This is convenient because the other day Gareth asked me to verify a behavior in different CHT Android versions, so I just had a small Apdex settings file for the default config. I changed the file path, skipped the login step (with the new settings), and tested it super fast in the 3 APK versions he requested.

File: tests/performance/apdex-score/sample-settings-files/default-config-navigation-test.json
Comment (latin-panda) [inline tests/performance/apdex-score/sample-settings-files/default-config-navigation-test.json:1]: @ralfudx @tatilepizs I made this file for the default config when I got the PouchDB 9 assignment. 
It only tests the navigation and search.
 
We can only push configs that CHT Core repo has (demo, default) but not from partners' configs. 

Because running these tests can take hours and these files can get long, I recommend you plan and create other files (like test suites), for example: testing forms, testing initial replication 20 times, and testing by relaunching the app in strategic moments. So to best test the performance of CHT. [I see Rafa made some recommendations about this last week.](https://medic.slack.com/archives/C02GME3BQE6/p1721141152372509?thread_ts=1721060268.713109&cid=C02GME3BQE6)

File: tests/performance/apdex-score/specs/test.apdex.js
Comment (latin-panda) [inline tests/performance/apdex-score/specs/test.apdex.js:1]: This was just a rename of `test.apdex.js` to `test.spec.js`, it follows the same we have with integration tests and others in the `/test/` folder

File: tests/performance/apdex-score/specs/test.spec.js
Comment (latin-panda) [inline tests/performance/apdex-score/specs/test.spec.js:17]: This is especially useful when used with `capabilities.noReset` settings, it speeds up the test a lot, especially when we are building the tests for a new instance. Because we don't need to wait to install the app, download 7000 docs, login, etc. So a faster start.

File: tests/performance/apdex-score/specs/test.spec.js
Comment (latin-panda) [inline tests/performance/apdex-score/specs/test.spec.js:65]: Because we can skip any test case now, we can have smaller suites or run specific cases easily.
I have separated the task load and task submit cases to allow testing navigation only, and to follow the report and contact forms way.

File: tests/performance/apdex-score/README.md
Comment (latin-panda) [inline tests/performance/apdex-score/README.md:1]: @ralfudx @tatilepizs I updated the Readme file with the new changes. 

@ralfudx, the setup steps for our community members are missing, for many of them this is a new technology, and they need high detail on how to do Apdex successfully. Please complete the documentation [in the main PR ](https://github.com/medic/cht-core/pull/9037), once it's approved by the reviewers, just copy and paste it into cht-docs repo's [automation test section. ](https://docs.communityhealthtoolkit.org/contribute/code/core/automated-tests/)

File: tests/page-objects/apdex/page.js
Comment (latin-panda) [reply tests/page-objects/apdex/page.js:201]: Thanks for checking @ralfudx. I agree that it is better not to use the index for finding this element

File: tests/performance/apdex-score/README.md
Comment (tatilepizs) [inline tests/performance/apdex-score/README.md:189]: I am a little confused with this section.
For the `"selector":` do I need to add selectors for all the fields available in the form? and the same for the `"dropdownOption":`, should I add the value of any dropdown in the form? Or should I add selectors for all the dropdowns available in the form?

**Review by tatilepizs — CHANGES_REQUESTED** [review]
Thank you for all the improvements @latin-panda ⭐ 
I left some comments/questions

File: tests/performance/apdex-score/README.md
Comment (latin-panda) [reply tests/performance/apdex-score/README.md:189]: This is inside `fields`, which is an array. Each element of the array is about 1 field. 
The "selector" is to click on that field and set focus: a radio, input text, dropdown, etc.
The "dropdownOption" is to click on the dropdown's option we want. It's better explained in the `Fields` section below L313. I missed updating it in this sample json, fixed!

For example, this is from MoH Kenya config form:
```
{
  "id": "county_residence",
  "selector": "//android.view.View[contains(@text, \"County of residence\")]",
   "dropdownOption": "//android.widget.CheckedTextView[@resource-id=\"android:id/text1\" and @text=\"BARINGO\"]"
},
```
It will click on the field `County of residence`, which will open the dropdown, then it will click on `BARINGO` option. 

The `selector` behaves the same across the whole settings file. It'll click on the selectors. This is convenient when you want to click on a label to remove the focus on a field, but this is usually not needed. I had to do it only once (Enketo being Enketo). 

There are some videos about[ this in here](https://drive.google.com/drive/folders/1d19YD3D2nxQkEorsd4QTnDK1UOPHS31U)

I think once we have more settings file examples, then we can remove this json from the REAME.

File: tests/page-objects/apdex/page.js
Comment (latin-panda) [reply tests/page-objects/apdex/page.js]: It worked for most. 
There is one that I had to create new function to click on links

File: tests/page-objects/apdex/page.js
Comment (latin-panda) [reply tests/page-objects/apdex/page.js]: I used `//android.widget.TextView`

**Comment by latin-panda** [comment]
@ralfudx @tatilepizs is ready for review again

@ralfudx can you please run with one of your settings files? It be good to have a second pair of eyes on that

**Comment by latin-panda** [comment]
Hi @ralfudx how did it go? Is it good?

File: tests/performance/apdex-score/README.md
Comment (tatilepizs) [reply tests/performance/apdex-score/README.md]: 😅  Thank you very much for the explanation Jenny!

**Review by tatilepizs — APPROVED** [review]
Looks really good @latin-panda thank you for all the improvements 🌟

**Pull request merged**
